Mathlib/AlgebraicGeometry/Morphisms/Separated.lean

```diff
@@ -90,6 +92,63 @@ instance (priority := 100) [h : IsAffineHom f] : IsSeparated f := by
   rw [MorphismProperty.arrow_mk_iso_iff @IsSeparated (arrowIsoSpecΓOfIsAffine f)]
   infer_instance
 
+instance {S T : Scheme.{u}} (f : X ⟶ S) (g : Y ⟶ S) (i : S ⟶ T) [IsSeparated i] :
+    IsClosedImmersion (pullback.mapDesc f g i) :=
+  IsClosedImmersion.stableUnderBaseChange (pullback_map_diagonal_isPullback f g i)
+    inferInstance
+
+instance [IsSeparated g] :
+    IsClosedImmersion (pullback.lift (𝟙 _) f (Category.id_comp (f ≫ g))) := by
+  rw [← MorphismProperty.cancel_left_of_respectsIso @IsClosedImmersion (pullback.fst f (𝟙 Y))]
+  rw [← MorphismProperty.cancel_right_of_respectsIso @IsClosedImmersion _
+    (pullback.congrHom rfl (Category.id_comp g)).inv]
+  convert (inferInstanceAs <| IsClosedImmersion (pullback.mapDesc f (𝟙 _) g)) using 1
+  ext : 1 <;> simp [pullback.condition]
+
 end IsSeparated
 
+lemma IsClosedImmersion.of_comp [IsClosedImmersion (f ≫ g)] [IsSeparated g] :
+    IsClosedImmersion f := by
+  rw [← pullback.lift_snd (𝟙 _) f (Category.id_comp (f ≫ g))]
+  have := IsClosedImmersion.stableUnderBaseChange.snd (f ≫ g) g inferInstance
+  infer_instance
+
+lemma IsSeparated.of_comp [IsSeparated (f ≫ g)] : IsSeparated f := by
+  have := IsSeparated.diagonal_isClosedImmersion (f := f ≫ g)
+  rw [pullback.diagonal_comp] at this
+  exact ⟨@IsClosedImmersion.of_comp _ _ _ _ _ this inferInstance⟩
+
+lemma IsSeparated.iff_of_comp [IsSeparated g] : IsSeparated (f ≫ g) ↔ IsSeparated f :=
+  ⟨fun _ ↦ .of_comp f g, fun _ ↦ inferInstance⟩
+
+namespace Scheme
+
+/-- A scheme `X` is separated if the diagonal map `X ⟶ X × X` is a closed immersion. -/
+@[mk_iff]
+protected class IsSeparated (X : Scheme.{u}) : Prop where
+  isClosedImmersion_prod_lift : IsClosedImmersion (prod.lift (𝟙 X) (𝟙 X))
```

Comment:
I think the definition should be `IsSeparated (terminal.from X)` instead as this seems to be the standard definition in the literature (see e.g. https://stacks.math.columbia.edu/tag/01KK or EGA I 5.4.1). In any case could you please point here to `isSeparated_iff_isSeparated_terminal`?